### DISCUSSION TITLE
detect last rows batch with Cursor.fetchmany query

### DISCUSSION THREAD

**Discussion by eadmaster** [oracle/python-oracledb#194] [Q&A]
Here i'd like to detect if the currently-processed rows batch is the last one **before** the final call to `cursor.fetchmany`.

My code:
````
cursor = connection.cursor()
cursor.execute("select * from MyTable")
num_rows = 10
objs_to_store = []
while True:
    rows = cursor.fetchmany(size=num_rows)
    if not rows:
        break
    for row in rows:
        # process row, generate curr_obj
        objs_to_store.append(curr_obj)
        if len(objs_to_store)==self.DJANGO_MAX_BULK_SIZE or is_last_row_batch:  # <--- also need to check if it is the last rows batch here
             # perform bulk insert

````

**Comment by cjbj** [comment]
If a fetchmany(size=X) call returns X rows, you **have** to do another fetchmany() call to see if there are any more rows to be returned from the DB since python-oracledb doesn't have any EOF information until all rows have been fetched.

**Comment by eadmaster** [comment]
what about this then?

````
cursor = connection.cursor()
cursor.execute("select * from MyTable")
num_rows = 10
objs_to_store = []
while True:
    rows = cursor.fetchmany(size=num_rows)
    if not rows:
        break
    for row in rows:
        # process row, generate curr_obj
        objs_to_store.append(curr_obj)
        if len(objs_to_store)==self.DJANGO_MAX_BULK_SIZE or ( len(rows)<num_rows and row==rows[-1]):  # reached DJANGO_MAX_BULK_SIZE or last rows batch
             # perform bulk insert
             objs_to_store = []
````

after fetching the last batch, i expect `len(rows)<num_rows` to be true:

>  If the number of rows available to be fetched is fewer than the amount requested, fewer rows will be returned.

([source](https://cx-oracle.readthedocs.io/en/latest/api_manual/cursor.html#Cursor.fetchmany))
